### ISSUE TITLE
Need to support multiple "internal states" in a GridComp

### ISSUE THREAD

**Issue by tclune** [esmf-org/esmf#330]
In GEOS/MAPL we canonically have 3 "internal" states for each user grid comp:
1.  wrapping an ordinary ESMF_State that is used to hold fields that are not part of import/export.  
2. wrapping a derived type used by the MAPL framework to provide additional services to a user gridcomp
3. occasionally to wrap a user-defined derived type that encapsulates various things such as values retrieved from the resource file (avoid repeated I/O) but can also be various things that don't have good ESMF analogs.   All private to the component.

For many years we have been using the undocumented interface `ESMF_UserCompGetInternalState` (and Set) and are trying to "do better".  There is a similar interface that is supported:  `ESMF_GridCompGetInternalState` (and Set), but this version only allows a single internal state.    The undocumented one has an extra string argument that is used to name each internal state.

We would like to either elevate the undocumented interface to be part of the public API **or** have an optional string argument on the existing public interface that provides the additional functionality.

See also:  https://github.com/GEOS-ESM/MAPL/issues/3078

**Comment by anntsay** [comment]
Bob:  already talked about this on the NASA call. NASA is doing these in private interface and requesting to have this public... Bob will take a look at the request to see if this could be part of effort for his 2019 funding.

**Comment by theurich** [comment]
@tclune at your convenience, please take a look at the new InternalState section in the RefDoc: https://earthsystemmodeling.org/docs/nightly/feature/internal-state-cleanup/ESMF_refdoc/node4.html#SECTION04080000000000000000. The InternalState API replaces the old `GridComp`, `CplComp`, and `UserComp` entry points that were there before, and will be deprecated starting with 8.9.0.

**Comment by tclune** [comment]
@theurich Looks good.  Thank you.

**Comment by mathomp4** [comment]
> [@mathomp4](https://github.com/mathomp4) when you get a chance, could you run discover/nag on the `feature/internal-state-cleanup` branch? I am pretty sure it will have failures... which will be the baseline before addressing the deferred type issue on NAG before the merge.

@theurich Sorry just saw this. I'll set one off tomorrow when I'm at work. (I'd do it now but Apple is updating my phone, so no token for a bit...)

**Comment by mathomp4** [comment]
Okay. The jobs are submitted. Hopefully they get through the queue at a good pace.

**Comment by theurich** [comment]
> Okay. The jobs are submitted. Hopefully they get through the queue at a good pace.

@mathomp4 any update on these jobs? Thanks.

**Comment by mathomp4** [comment]
Well that's weird. I submitted them last week. I wonder why they didn't get pushed up?

**Comment by mathomp4** [comment]
Okay. I am really confused. I swear I am submitting these. Do I need to maybe clear out some cache or something?

**Comment by oehmke** [comment]
Hi Matt, How are you submitting them?

**Comment by mathomp4** [comment]
I just ran this:

```
> /discover/nobackup/projects/gmao/SIteam/ESMF_Testing/esmf-test-scripts/python_scripts/test_esmf.py -r $ESMF_TEST_ROOT -m discover --branch feature/internal-state-cleanup --filter 15,16
```
and it said:
```
INFO: Setting up test case: [nag 7.2.15 / openmpi nag_7.2.7215-gcc_12.3.0 / None / O / feature/internal-state-cleanup]
INFO: Setting up test case: [nag 7.2.15 / openmpi nag_7.2.7215-gcc_12.3.0 / None / g / feature/internal-state-cleanup]
INFO: Submitted case: [nag 7.2.15 / openmpi nag_7.2.7215-gcc_12.3.0 / None / O / feature/internal-state-cleanup]
INFO: Submitted case: [nag 7.2.15 / openmpi nag_7.2.7215-gcc_12.3.0 / None / g / feature/internal-state-cleanup]
```

**Comment by theurich** [comment]
@mathomp4 I took a look under your run directories, and I see that those tests ran. However, I have no idea why the results aren't pushed to the artifacts repo. But I don't think we need to worry about that right now.

So I saw what I was expecting with branch `feature/internal-state-cleanup` on discover/nag. Basically the InternalState feature that depends on deferred type arguments (`type(*)`) showed the expected failures (because support for that feature was disabled for NAG). I just enabled it on `feature/internal-state-cleanup` for testing. Please kick off discover/nag on that branch one more time... I'll just monitor your test directories directly. Thanks.

**Comment by mathomp4** [comment]
@theurich Well this is a...coincidence. I recently installed NAG 7.2.28 on discover (it fixes a weird exponentially slowing build issue we found). And to celebrate/test, I just kicked off a build of that branch (mainly because I had the command in my history). I did also try and be fancy and added `netcdf` to the build as well.

So:
```
INFO: Setting up test case: [nag 7.2.28 / openmpi nag_7.2.7228-gcc_12.3.0 / other/esmf-netcdf/nagfor_7.2-7228-openmpi_4.1.6 / O / feature/internal-state-cleanup]
INFO: Setting up test case: [nag 7.2.28 / openmpi nag_7.2.7228-gcc_12.3.0 / other/esmf-netcdf/nagfor_7.2-7228-openmpi_4.1.6 / g / feature/internal-state-cleanup]
INFO: Submitted case: [nag 7.2.28 / openmpi nag_7.2.7228-gcc_12.3.0 / other/esmf-netcdf/nagfor_7.2-7228-openmpi_4.1.6 / O / feature/internal-state-cleanup]
INFO: Submitted case: [nag 7.2.28 / openmpi nag_7.2.7228-gcc_12.3.0 / other/esmf-netcdf/nagfor_7.2-7228-openmpi_4.1.6 / g / feature/internal-state-cleanup]
```
Of course, if I got something wrong in the config, the build will fail.

**Comment by theurich** [comment]
@mathomp4 I agree, this looks good. Thanks for updating to NAG 7.2.28 on discover, and for running the branch tests.

One strange issue (unrelated to the internal state work) seems to be that with NAG 7.2.28 we now get a SEGV in one of the NUOPC app proto tests. Also happens on develop, so it doesn't seem related to the feature branch here. I'll open an issue to investigate.

**Comment by theurich** [comment]
Resolved with #383.
